flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -249,10 +248,28 @@ private void interruptSourceThread(boolean interrupt) {
     // ------------------------------------------------------------------------
 
     @Override
-    public Future<Boolean> triggerCheckpointAsync(
+    public CompletableFuture<Boolean> triggerCheckpointAsync(
             CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
         if (!externallyInducedCheckpoints) {
-            return super.triggerCheckpointAsync(checkpointMetaData, checkpointOptions);
+            if (checkpointOptions.getCheckpointType().shouldDrain()) {
+                mainMailboxExecutor.execute(
+                        () -> {
+                            setSynchronousSavepoint(checkpointMetaData.getCheckpointId(), true);
+                            wasDrained = true;
+                            if (mainOperator != null) {
+                                mainOperator.stop();
+                            }
+                        },
+                        "stop legacy source and set synchronous savepoint");
+                return sourceThread
+                        .getCompletionFuture()
+                        .thenCompose(
```

Comment:
the change from `Future` to `CompletableFuture` in `triggerCheckpointAsync` is because of this compose?

Reply:
Correct. Note, it is a change for the StreamTask. I did not change it in the `AbstractInvokable` interface.